File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccessTesting/FakeVaultSession.swift
import Foundation
import SecretaryVaultAccess

/// In-memory `VaultSession` for host tests. `recordsByBlock` is mutable so the
/// write methods model real add/edit/delete/restore state transitions. Field
/// `reveal` closures capture the stored plaintext.
public final class FakeVaultSession: VaultSession {
    public let vaultUuidHex: String
    private var blocks: [BlockSummary]
    private var recordsByBlock: [[UInt8]: [RecordView]]
    private var nextUuidByte: UInt8 = 0xA0
    public private(set) var readCount = 0
    public private(set) var wipeCount = 0
    /// Spy: the includeDeleted value passed to the most recent readBlock.
    public private(set) var lastIncludeDeleted = false
    /// Test seam: when set, the NEXT create/rename/move throws this once, then clears.
    public var failNextWrite: VaultAccessError?

    public init(vaultUuidHex: String,
                blocks: [BlockSummary],
                recordsByBlock: [[UInt8]: [RecordView]]) {
        self.vaultUuidHex = vaultUuidHex
        self.blocks = blocks
        self.recordsByBlock = recordsByBlock
    }

    public func blockSummaries() -> [BlockSummary] { blocks }

    public func readBlock(blockUuid: [UInt8], includeDeleted: Bool) throws -> [RecordView] {
        readCount += 1
        lastIncludeDeleted = includeDeleted
        guard let records = recordsByBlock[blockUuid] else {
            throw VaultAccessError.blockNotFound(hex(blockUuid))
        }
        // Model the Rust gate: withhold tombstoned records unless asked.
        return includeDeleted ? records : records.filter { !$0.tombstone }
    }

    @discardableResult
    public func appendRecord(blockUuid: [UInt8], content: RecordContentInput) throws -> [UInt8] {
        try requireBlock(blockUuid)
        let uuid = mintUuid()
        recordsByBlock[blockUuid]?.append(recordView(uuid: uuid, content: content, tombstone: false))
        return uuid
    }

    public func editRecord(blockUuid: [UInt8], recordUuid: [UInt8], content: RecordContentInput) throws {
        let idx = try liveIndex(blockUuid, recordUuid)
        recordsByBlock[blockUuid]?[idx] = recordView(uuid: recordUuid, content: content, tombstone: false)
    }

    public func tombstoneRecord(blockUuid: [UInt8], recordUuid: [UInt8]) throws {
        let idx = try liveIndex(blockUuid, recordUuid)
        setTombstone(blockUuid, idx, true)
    }

    public func resurrectRecord(blockUuid: [UInt8], recordUuid: [UInt8]) throws {
        try requireBlock(blockUuid)
        guard let idx = recordsByBlock[blockUuid]?.firstIndex(
            where: { $0.uuid == recordUuid && $0.tombstone }) else {
            throw VaultAccessError.recordNotFound(hex(recordUuid))
        }
        setTombstone(blockUuid, idx, false)
    }

    @discardableResult
    public func createBlock(blockName: String) throws -> [UInt8] {
        try throwIfInjected()
        let uuid = mintUuid()
        blocks.append(BlockSummary(uuid: uuid, name: blockName, createdAtMs: 0, lastModMs: 0))
        recordsByBlock[uuid] = []
        return uuid
    }

    public func renameBlock(blockUuid: [UInt8], newName: String) throws {
        try throwIfInjected()
        guard let idx = blocks.firstIndex(where: { $0.uuid == blockUuid }) else {
            throw VaultAccessError.blockNotFound(hex(blockUuid))
        }
        let old = blocks[idx]
        blocks[idx] = BlockSummary(uuid: old.uuid, name: newName,
                                   createdAtMs: old.createdAtMs, lastModMs: old.lastModMs)
    }

    @discardableResult
    public func moveRecord(sourceBlockUuid: [UInt8], targetBlockUuid: [UInt8],
                           sourceRecordUuid: [UInt8]) throws -> [UInt8] {
        try throwIfInjected()
        let idx = try liveIndex(sourceBlockUuid, sourceRecordUuid)
        try requireBlock(targetBlockUuid)
        guard let src = recordsByBlock[sourceBlockUuid]?[idx] else {
            throw VaultAccessError.recordNotFound(hex(sourceRecordUuid))
        }
        let newUuid = mintUuid()
        // copy-before-delete: land the copy in the target, THEN tombstone the source.
        recordsByBlock[targetBlockUuid]?.append(RecordView(
            uuid: newUuid, type: src.type, tags: src.tags, fields: src.fields, tombstone: false))
        setTombstone(sourceBlockUuid, idx, true)
        return newUuid
    }

    public func wipe() { wipeCount += 1 }

    // MARK: - helpers

    private func throwIfInjected() throws {
        if let e = failNextWrite { failNextWrite = nil; throw e }
    }

    private func requireBlock(_ blockUuid: [UInt8]) throws {
        guard recordsByBlock[blockUuid] != nil else {
            throw VaultAccessError.blockNotFound(hex(blockUuid))
        }
    }

    private func liveIndex(_ blockUuid: [UInt8], _ recordUuid: [UInt8]) throws -> Int {
        try requireBlock(blockUuid)
        guard let idx = recordsByBlock[blockUuid]?.firstIndex(
            where: { $0.uuid == recordUuid && !$0.tombstone }) else {
            throw VaultAccessError.recordNotFound(hex(recordUuid))
        }
        return idx
    }

    private func setTombstone(_ blockUuid: [UInt8], _ idx: Int, _ value: Bool) {
        guard let old = recordsByBlock[blockUuid]?[idx] else { return }
        recordsByBlock[blockUuid]?[idx] = RecordView(
            uuid: old.uuid, type: old.type, tags: old.tags, fields: old.fields, tombstone: value)
    }

    private func recordView(uuid: [UInt8], content: RecordContentInput, tombstone: Bool) -> RecordView {
        let fields = content.fields.map { f -> FieldView in
            switch f.value {
            case .text(let s):
                return FieldView(name: f.name, kind: .text) { .text(s) }
            case .bytes(let b):
                return FieldView(name: f.name, kind: .bytes) { .bytes(b) }
            }
        }
        return RecordView(uuid: uuid, type: content.recordType, tags: content.tags,
                          fields: fields, tombstone: tombstone)
    }

    private func mintUuid() -> [UInt8] {
        let b = nextUuidByte
        nextUuidByte = nextUuidByte &+ 1
        return [UInt8](repeating: b, count: 16)
    }

    private func hex(_ bytes: [UInt8]) -> String { bytes.map { String(format: "%02x", $0) }.joined() }
}

File: android/browse-ui/src/androidTest/kotlin/org/secretary/browse/FakeVaultSession.kt
package org.secretary.browse

import kotlinx.coroutines.CompletableDeferred

/**
 * Instrumented-source VaultSession double (androidTest can't see the unit-test fake).
 *
 * Deliberately minimal subset of the host `:vault-access` `FakeVaultSession`: covers the
 * write gate and happy-path recording only. It does NOT include the `writeError` /
 * `rawWriteThrowable` error-injection fields — those are exercised in the host unit tests
 * where coroutine-test utilities are available. Do not add error-injection here; keep
 * this double focused on gate-based concurrency and happy-path instrumented scenarios.
 */
class FakeVaultSession(
    private val vaultUuidHex: String,
    blocks: List<BlockSummaryView>,
    recordsByBlockHex: Map<String, List<RecordSummaryView>> = emptyMap(),
    private val writeGate: CompletableDeferred<Unit>? = null,
) : VaultSession {
    var wiped: Boolean = false
        private set

    val appended: MutableList<Pair<String, RecordContentInput>> = mutableListOf()
    val edited: MutableList<Triple<String, String, RecordContentInput>> = mutableListOf()
    val created: MutableList<String> = mutableListOf()
    val renamed: MutableList<Pair<String, String>> = mutableListOf()         // blockHex -> newName
    val moved: MutableList<Triple<String, String, String>> = mutableListOf() // srcHex, tgtHex, recHex
    private var nextFakeUuidByte: Int = 0xA0
    private var nextFakeBlockByte: Int = 0xB0

    private val mutableBlocks: MutableList<BlockSummaryView> = blocks.toMutableList()
    private val records: MutableMap<String, MutableList<RecordSummaryView>> =
        recordsByBlockHex.mapValues { it.value.toMutableList() }.toMutableMap()

    override fun vaultUuidHex(): String = vaultUuidHex
    override fun blockSummaries(): List<BlockSummaryView> = mutableBlocks.toList()
    override suspend fun readBlock(blockUuid: ByteArray, includeDeleted: Boolean): List<RecordSummaryView> {
        val all = records[hexOfBytes(blockUuid)] ?: return emptyList()
        return if (includeDeleted) all.toList() else all.filter { !it.tombstone }
    }
    override suspend fun tombstoneRecord(blockUuid: ByteArray, recordUuid: ByteArray) {
        writeGate?.await()
        flip(hexOfBytes(blockUuid), hexOfBytes(recordUuid), tombstone = true)
    }
    override suspend fun resurrectRecord(blockUuid: ByteArray, recordUuid: ByteArray) {
        writeGate?.await()
        flip(hexOfBytes(blockUuid), hexOfBytes(recordUuid), tombstone = false)
    }
    override fun wipe() { wiped = true }

    override suspend fun appendRecord(blockUuid: ByteArray, content: RecordContentInput): ByteArray {
        writeGate?.await()
        val blockHex = hexOfBytes(blockUuid)
        appended += blockHex to content
        val uuid = ByteArray(16).also { it[15] = (nextFakeUuidByte and 0xff).toByte() }
        nextFakeUuidByte += 1
        val list = records.getOrPut(blockHex) { mutableListOf() }
        list += RecordSummaryView(
            uuidHex = hexOfBytes(uuid),
            type = content.recordType,
            tags = content.tags,
            createdAtMs = 0u,
            lastModMs = 0u,
            tombstone = false,
            fields = content.fields.map { it.toRevealableField() },
        )
        return uuid
    }

    override suspend fun editRecord(blockUuid: ByteArray, recordUuid: ByteArray, content: RecordContentInput) {
        writeGate?.await()
        val blockHex = hexOfBytes(blockUuid)
        val recordHex = hexOfBytes(recordUuid)
        edited += Triple(blockHex, recordHex, content)
        val list = records[blockHex]
        val i = list?.indexOfFirst { it.uuidHex == recordHex } ?: -1
        if (i < 0) throw VaultBrowseError.RecordNotFound(recordHex)
        list!![i] = list[i].copy(
            type = content.recordType,
            tags = content.tags,
            fields = content.fields.map { it.toRevealableField() },
        )
    }

    override suspend fun createBlock(blockName: String): ByteArray {
        writeGate?.await()
        val uuid = ByteArray(16).also { it[15] = (nextFakeBlockByte and 0xff).toByte() }
        nextFakeBlockByte += 1
        created += blockName
        mutableBlocks += BlockSummaryView(uuid, blockName, 0u, 0u)
        records.getOrPut(hexOfBytes(uuid)) { mutableListOf() }
        return uuid
    }

    override suspend fun renameBlock(blockUuid: ByteArray, newName: String) {
        writeGate?.await()
        val hex = hexOfBytes(blockUuid)
        val i = mutableBlocks.indexOfFirst { it.uuidHex == hex }
        if (i < 0) throw VaultBrowseError.BlockNotFound(hex)
        renamed += hex to newName
        val b = mutableBlocks[i]
        mutableBlocks[i] = BlockSummaryView(b.uuid, newName, b.createdAtMs, b.lastModifiedMs)
    }

    override suspend fun moveRecord(
        sourceBlockUuid: ByteArray,
        targetBlockUuid: ByteArray,
        sourceRecordUuid: ByteArray,
    ): ByteArray {
        writeGate?.await()
        val srcHex = hexOfBytes(sourceBlockUuid)
        val tgtHex = hexOfBytes(targetBlockUuid)
        val recHex = hexOfBytes(sourceRecordUuid)
        moved += Triple(srcHex, tgtHex, recHex)
        val srcList = records[srcHex] ?: throw VaultBrowseError.BlockNotFound(srcHex)
        val si = srcList.indexOfFirst { it.uuidHex == recHex && !it.tombstone }
        if (si < 0) throw VaultBrowseError.RecordNotFound(recHex)
        val moving = srcList[si]
        val newUuid = ByteArray(16).also { it[15] = (nextFakeUuidByte and 0xff).toByte() }
        nextFakeUuidByte += 1
        records.getOrPut(tgtHex) { mutableListOf() } += moving.copy(uuidHex = hexOfBytes(newUuid))
        srcList[si] = moving.copy(tombstone = true)
        return newUuid
    }

    private fun flip(blockHex: String, recordHex: String, tombstone: Boolean) {
        val list = records[blockHex] ?: return
        val i = list.indexOfFirst { it.uuidHex == recordHex }
        if (i >= 0) list[i] = list[i].copy(tombstone = tombstone)
    }
}

fun textField(name: String, value: String): RevealableField =
    RevealableField(name, FieldKind.Text) { RevealedValue.Text(value) }

private fun FieldContentInput.toRevealableField(): RevealableField = when (val v = value) {
    is FieldContentValue.Text -> RevealableField(name, FieldKind.Text) { RevealedValue.Text(v.value) }
    is FieldContentValue.Bytes -> RevealableField(name, FieldKind.Bytes) { RevealedValue.Bytes(v.value) }
}
